### PR TITLE
Clean up GUI tests related classes #559

### LINKED ISSUE #559: Clean up GUI tests related classes
`PersonListPanelHandle` and `PersonCardHandle` are classes that point to their respective nodes (`PersonListPanelHandle` points to the `PersonListPanel`, while `PersonCardHandle` points to `PersonCard`). The purpose of these handles is to encapsulate these nodes by exposing only certain operations that are required by tests.

- [x] Move `PersonListPanelHandle#isListMatching(ReadOnlyPerson...)` as this is somewhat like a test verification method. `PersonListPanelHandle` should have a method that returns a defensive copy of the internal list of `PersonListPanel`.
- [x] Move `PersonCardHandle#isSamePerson(ReadOnlyPerson)` for the same reason above.
- [x] I think that we should not return `Optional` for `getHandleToSelectedCard()`, since when we call a test method, we are expecting it to return us a value. As such, if there's no value, the method should throw an appropriate RuntimeException.
-------
`AddressBookGuiTest` is a class that's attempting to store everything that the subclasses will use. I think that it's purpose is to set up the environment for subclasses (like setting up the stage etc) so that the code does not have to be repeated across all the tests. As such:

- [x] Move test verification methods such as `assertCardMatchesPerson(PersonCardHandle, ReadOnlyPerson)` as these ought to belong to test utility classes.
- [x] Move other random methods like `raise(BaseEvent)`, which ought to belong to test utility classes as well.
- [x] `EventsCenter.clearSubscribers();` is called in `setup()` which is annotated as `@Before`. It's more intuitive to shift it to a method that is annotated `@After`, as clearing the subscribers seem like a tear down method rather than a set up method.
-------
- [x] `GuiRobot`'s access modifiers are wrong. (PAUSE_FOR_HUMAN_DELAY_MILLISECONDS, DEFAULT_WAIT_FOR_EVENT_TIMEOUT_MILLISECONDS, waitForEvent)

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #559

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/8] [GuiRobot: Update visibility modifiers of constants](https://github.com/se-edu/addressbook-level4/pull/617/commits/199cb68b9fb6e97f0fc1a5ac12a18a975a808980)
* [v1 2/8] [AddressBookGuiTest: Move assertListSize\(int\) to GuiTestAssert](https://github.com/se-edu/addressbook-level4/pull/617/commits/c591b380c6b2021dbccd3d733a8960e53e451241)
* [v1 3/8] [AddressBookGuiTest: Move assertResultMessage\(String\) to GuiTestAssert](https://github.com/se-edu/addressbook-level4/pull/617/commits/57a79658ac45f659c32e85da0b59c2bbf906de69)
* [v1 4/8] [AddressBookGuiTest: Move raise\(BaseEvent\) to EventsUtil](https://github.com/se-edu/addressbook-level4/pull/617/commits/5680e99d121cd7b336558611247e2f58f702156b)
* [v1 5/8] [EventsUtil: Update method names](https://github.com/se-edu/addressbook-level4/pull/617/commits/b4d72d108a63829d13e2a0f8d1808692320fbaa6)
* [v1 6/8] [PersonListPanelHandle: Remove unused method](https://github.com/se-edu/addressbook-level4/pull/617/commits/f3a7ac6da1474c7a784ab8da0fb074d440218438)
* [v1 7/8] [PersonListPanelHandle: Add isCardSelected\(\)](https://github.com/se-edu/addressbook-level4/pull/617/commits/979dedbf8a9e5906288a2c19a3cf9aad643a294a)
* [v1 8/8] [PersonListPanelHandle: Update return type of getHandleToSelectedCard\(\)](https://github.com/se-edu/addressbook-level4/pull/617/commits/f44a101b6a6b0dea8a10a7cd0e28b2511803c760)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/816f07f953e8855556176cc0b6f9ca8a13a7b0f7...f44a101b6a6b0dea8a10a7cd0e28b2511803c760))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/617/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/guitests/guihandles/PersonListPanelHandle.java [hunk 1]
```diff
@@ -36,6 +36,19 @@ public PersonListPanelHandle(ListView<PersonCard> personListPanelNode) {
     }
 
     /**
+     * Returns true if a card is currently selected.
+     */
+    public boolean isCardSelected() {
```
Comment (yamgent) [inline src/test/java/guitests/guihandles/PersonListPanelHandle.java]: `isAnyCardSelected()`

File: src/test/java/guitests/guihandles/PersonListPanelHandle.java [hunk 2]
```diff
@@ -25,14 +25,14 @@ public PersonListPanelHandle(ListView<PersonCard> personListPanelNode) {
      * Returns a handle to the selected {@code PersonCardHandle}.
      * A maximum of 1 item can be selected at any time.
```
Comment (yamgent) [inline src/test/java/guitests/guihandles/PersonListPanelHandle.java]: > Since it is a test method, we expect callers to only call it when a card
> is selected. As such, we do not need to account for the case when no
> cards are selected.


Mention this pre-condition in the header comments.

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/8] [GuiRobot: Update visibility modifiers of constants](https://github.com/se-edu/addressbook-level4/pull/617/commits/199cb68b9fb6e97f0fc1a5ac12a18a975a808980)
* [v2 2/8] [AddressBookGuiTest: Move assertListSize\(int\) to GuiTestAssert](https://github.com/se-edu/addressbook-level4/pull/617/commits/c591b380c6b2021dbccd3d733a8960e53e451241)
* [v2 3/8] [AddressBookGuiTest: Move assertResultMessage\(String\) to GuiTestAssert](https://github.com/se-edu/addressbook-level4/pull/617/commits/57a79658ac45f659c32e85da0b59c2bbf906de69)
* [v2 4/8] [AddressBookGuiTest: Move raise\(BaseEvent\) to EventsUtil](https://github.com/se-edu/addressbook-level4/pull/617/commits/5680e99d121cd7b336558611247e2f58f702156b)
* [v2 5/8] [EventsUtil: Update method names](https://github.com/se-edu/addressbook-level4/pull/617/commits/b4d72d108a63829d13e2a0f8d1808692320fbaa6)
* [v2 6/8] [PersonListPanelHandle: Remove unused method](https://github.com/se-edu/addressbook-level4/pull/617/commits/f3a7ac6da1474c7a784ab8da0fb074d440218438)
* [v2 7/8] [PersonListPanelHandle: Add isAnyCardSelected\(\)](https://github.com/se-edu/addressbook-level4/pull/617/commits/dac545de6d73a0f54cc5f7c29c3cf25f3714b24b)
* [v2 8/8] [PersonListPanelHandle: Update return type of getHandleToSelectedCard\(\)](https://github.com/se-edu/addressbook-level4/pull/617/commits/5813b67c6f3931ac31bfdd6b7de4350572e4b0cc)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/816f07f953e8855556176cc0b6f9ca8a13a7b0f7...5813b67c6f3931ac31bfdd6b7de4350572e4b0cc)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/d1dff5c0caf2a5a1632c377c3ac0a3e5f06db237~1...d1dff5c0caf2a5a1632c377c3ac0a3e5f06db237))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/617/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/testutil/EventsUtil.java [hunk 3]
```diff
@@ -12,7 +13,14 @@
      * Posts {@code event} to all registered subscribers. This method will return successfully after the {@code event}
      * has been posted to all subscribers.
      */
-    public static void post(BaseEvent event) {
+    public static void postAndWait(BaseEvent event) {
```
Comment (damithc) [inline src/test/java/seedu/address/testutil/EventsUtil.java]: `postNow` (to go with `postLater`)?

File: src/test/java/guitests/AddCommandTest.java [hunk 4]
```diff
@@ -37,7 +38,7 @@ public void add() throws Exception {
 
         //add duplicate person
         runCommand(PersonUtil.getAddCommand(HOON));
-        assertResultMessage(AddCommand.MESSAGE_DUPLICATE_PERSON);
+        assertResultMessage(getResultDisplay(), AddCommand.MESSAGE_DUPLICATE_PERSON);
```
Comment (damithc) [inline src/test/java/guitests/AddCommandTest.java]: Although not part of this PR, `getResultDisplay()` does not sound like it is returning a string. It sounds like returning a UI component.

Comment (Zhiyuan-Amos) [inline src/test/java/guitests/AddCommandTest.java]: It isn't returning a `String`, but a UI component :P 

The method signature is `assertResultMessage(ResultDisplayHandle, String)`.

Comment (damithc) [inline src/test/java/guitests/AddCommandTest.java]: I see. It's fine then :-p
